connection.go

```diff
@@ -1666,3 +1734,47 @@ func (conn *Connection) ServerProtocolInfo() ProtocolInfo {
 func (conn *Connection) ClientProtocolInfo() ProtocolInfo {
 	return clientProtocolInfo.Clone()
 }
+
+func shutdownEventCallback(event WatchEvent) {
+	// Receives "true" on server shutdown.
+	// See https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/graceful_shutdown/
+	// step 2.
+	val, ok := event.Value.(bool)
+	if ok && val {
+		event.Conn.processShutdown()
+	}
+}
+
+func (conn *Connection) processShutdown() {
+	// Forbid state changes.
+	conn.mutex.Lock()
+	defer conn.mutex.Unlock()
+
+	conn.state = connShutdown
+	conn.notify(Shutdown)
+
+	for conn.state == connConnected && atomic.LoadUint32(&(conn.requestCnt)) != 0 {
+		// Use cond var on conn.mutex since request execution may
+		// call reconnect(). It is ok if state changes as part of
+		// reconnect since Tarantool server won't allow to reconnect
+		// in the middle of shutting down.
+		conn.cond.Wait()
+	}
+
+	// Do not unregister task explicitly here since connection teardown
+	// has the same effect. To clean up connection resources,
+	// unregister on full close.
+
+	if conn.state == connConnected {
+		// Start to reconnect based on common rules, same as in net.box.
+		// Reconnect also closes the connection: server waits until all
+		// subscribed connections are terminated.
+		// See https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/graceful_shutdown/
+		// step 3.
+		go conn.reconnect(
+			ClientError{
+				ErrConnectionClosed,
+				"connection closed after server shutdown",
+			}, conn.c)
+	}
```

Comment:
To avoid shutdown after reconnect:
```suggestion
        c := conn.c
	for conn.state == connConnected && atomic.LoadUint32(&(conn.requestCnt)) != 0 && conn.c == c {
		// Use cond var on conn.mutex since request execution may
		// call reconnect(). It is ok if state changes as part of
		// reconnect since Tarantool server won't allow to reconnect
		// in the middle of shutting down.
		conn.cond.Wait()
	}

	// Do not unregister task explicitly here since connection teardown
	// has the same effect. To clean up connection resources,
	// unregister on full close.

	if conn.state == connConnected && conn.c == c {
		// Start to reconnect based on common rules, same as in net.box.
		// Reconnect also closes the connection: server waits until all
		// subscribed connections are terminated.
		// See https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/graceful_shutdown/
		// step 3.
		go conn.reconnect(
			ClientError{
				ErrConnectionClosed,
				"connection closed after server shutdown",
			}, conn.c)
	}
```

Reply:
Reworked